### PR TITLE
Feature/adapter refactor (#962, #963, #965, #1035, #1090)

### LINKED ISSUE #962: Clean up global connection pooling
## Feature
We should clean up global connection pooling and instead attach any connection pool to the adapter object itself.

### Feature description
We currently store the connection pool in a global and it should instead be part of the adapter itself. This should make a client/server model for dbt a lot more sustainable, as potentially a server would have multiple adapters, each with its own configuration and therefore connection pool.

This requires #961.

### Who will this benefit?
dbt developers

### LINKED ISSUE #963: Deprecate and remove adapter methods
## Feature
Deprecate and remove unused/undesirable adapter methods

### Feature description
This is a prerequisite for making adapters-as-plugins. This is in large part just a discussion we need to have where we hammer out what we need and what we don't need adapters to implement, define an abstract base class that adapters use (probably this is mostly DefaultAdapter), and make adapters derive from that class directly instead of the current common chain through the Postgres adapter. There are really two sides to the API that comes out of this:
1) A list of methods a custom third-party adapter would need to define to be usable by dbt
2) A list of methods dbt promises are available to materializations/macros/etc

We probably want to wait until at least #961 is implemented before we start on this, if only from a merge conflict management perspective.

### Who will this benefit?
dbt developers.

### LINKED ISSUE #965: Optional adapters and adapter dependencies
## Feature
dbt should provide a mechanism for users to not require dependencies for databases a user doesn't use.

### Feature description
An example of this is the Snowflake adapter, which has a number of painful dependency restrictions. While it's unfortunate that Snowflake users have to deal with them, there's no reason Redshift/Bigquery/Postgres users should have to, and the same goes for those other dependencies.
As a first pass, we really only need to make it possible to turn deps _off_, and it's ok and reasonable to default to requiring all dependencies.

This feature should also lay the groundwork for a real adapters-as-plugins architecture, since that's the logical next step.

This probably depends on #961. At least it would be wise to wait for that to do this, as they are not really compatible changes and re-doing the work sounds unpleasant.

### Who will this benefit?
dbt users who only care about one or a couple databases.

### LINKED ISSUE #1035: Clean up adapter inheritance hierarchy

### LINKED ISSUE #1090: Lazy Load adapters
## Feature
Lazy load adapter modules, only importing the adapter package once the adapter type has been chosen.

### Feature description
Background: Currently, all supported adapters are built into dbt, and their configuration requirements are built into the configuration specification at definition time. The branch for #965 changes that to start off with the configuration specification empty, and opted-in adapters are injected into the configuration specification (and the list of available adapter classes) at import time.

Lazy loading will defer that import even further, to configuration resolution time, and only import the adapter specified by the chosen `profile` and `target` combination. 

One side effect of this feature: the adapter's `type` will be required to be identical to (or at least easily and uniquely mapped to) the module name, so we can guess what import path the adapter lives at from its name alone. This seems like a very reasonable constraint, given that we already do it.

### Who will this benefit?
One immediate benefit is that this should improve import times (and therefore dbt startup times) without any user intervention. It will be most noticeable when doing very fast operations like `dbt debug` or `dbt --help`, `compile`, `run`, `seed`, etc. will probably see no benefit.

This will also make it easier for adapters as plugins on a couple fronts:
 - there's no more need to worry about adapter macros conflicting with one another if they end up in the `dbt` namespace (likely), since only one adapter plugin can be available at runtime
 - This solves the plugin discovery problem nicely

### Risks/concerns
I haven't looked into this a ton, but I suspect making integration tests support this will be the most difficult aspect. Lazy loading is fairly straightforward and mostly requires moving some configuration/validation code around, with slight tweaks to the existing adapter factory.

### PR DESCRIPTION
**Pull request by beckjake**
fixes #963, fixes #1035, fixes #962, fixes #965, fixes #1090

This merges PRs #1064 and #1091 into, I have combined them as a single description below.

### User-facing changes:
 - Remove deprecated methods
 - Connections are now managed on a per-adapter-instance basis. As adapters are generally singletons, that should be no different than before. This means that adapters have full control over their connection management,  which seems right.
 - All adapters are now optional, and installers have to explicitly opt-in with the [feature] syntax. 
 - Adapter packages load lazily, when the configuration's credentials object is parsed.

### Internal changes
 - Created a more consistent class hierarchy where bigquery adapter does not derive from postgres, snowflake and postgres have a common ancestor, etc
 - Defined what methods subclasses must override, use python's abstract classes to enforce them
 - Gave the DBWrapper its list of methods to expose via a decorator + metaclass model because I broke those arrays like 20 times during development and wow, what a pain to maintain with subclassing.
 - Split adapters into two components: A connection manager that doesn't understand relations/the manifest/anything about dbt, and the "real" adapter that does.
- `dbt.adapters.factory.ADAPTER_CLASSES` now starts off empty
-  When the chosen target references an adapter `typename`, dbt imports `dbt.adapters.typename` as `pkg`, registers `pkg.Adapter` for the type, and returns `pkg.Credentials` for validation. So packages need to expose their adapters/credentials under those names.
- Accordingly, contracts are validated at target parsing time, which occurs after regular configuration parsing
- Converting dbt internal schema representations into bigquery schema representations now happens in the bigquery adapter itself, purely for dependency reasons.
- adapter-specific Connection contracts are now part of the adapter packages themselves, instead of part of dbt.contracts.connection

### Test changes
I changed the tox environments (and so also the CI tests) to use the new `-e .[...]` syntax. Things seem to work.

### User impact

To get the existing dbt installation behavior, either continue to use `-r requirements.txt` or use `-e .[postgres,snowflake,redshift,bigquery]`. When this is on pypi, users will have to `pip install dbt[postgres,snowflake,redshift,bigquery]` to get the same behavior, which is unfortunate. Python does not really provide an opt-out mechanism for dependencies, only opt-in, and has no concept of a "default" features set.

If that is scary from a backwards-compatibility perspective, one gross idea I had was to make this something like `dbt-core`, and make a `dbt` package that depends upon `dbt-core[postgres,snowflake,redshift,bigquery]`.

#### Assorted thoughts

We also may want to add a deprecation warning to 0.12.x for `get_columns_in_table`, and then we could change this branch to remove it entirely. I've replaced it with `get_columns_in_relation`, which takes a `Relation`, and `get_columns_in_table` builds one and passes it in (like `drop`/`rename`/`truncate` did).

It wouldn't be too hard to just assign one connection per thread, as it is connections and threads are related (connection count is directly related to thread count) but disjoint (good luck mapping connections to threads or vice-versa). We could force that, by making the thread pool responsible for connections, so each thread gets one connection at start time. Then we would never have to track the existence of connections, just their state.

One thing this PR does not solve is the adapter-specific macros issue. Those are still included in all dbt runs all the time, and exist regardless of `[feature]` settings.

The connection management does leave me with this vague feeling that I should be able to do something like this:

```
with adapter.transaction() as conn:
   # connection created or reopened, begin called on __enter__
   conn.execute(...) # or add_query(), whatever
# commit/rollback + close called as appropriate on __exit__
# connection is now back in the pool, in a closed state
```

As it is I'm never really confident that all connections have been closed properly.

### CODE REVIEW

File: dbt/adapters/base/connections.py [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+    def close(cls, connection):
+        if dbt.flags.STRICT_MODE:
+            assert isinstance(connection, Connection)
+
+        # if the connection is in closed or init, there's nothing to do
+        if connection.state in {'closed', 'init'}:
+            return connection
+
+        # On windows, sometimes connection handles don't have a close() attr.
+        if hasattr(connection.handle, 'close'):
```
Comment (drewbanin) [inline dbt/adapters/base/connections.py]: it seems weird to me that we wouldn't call `close()` on the handle on Windows. Is this right?

File: dbt/adapters/base/impl.py [hunk 2]
```diff
@@ -0,0 +1,705 @@
...
+
+    def __init__(self, config):
+        self.config = config
+        self.cache = RelationsCache()
+        self.connections = self.ConnectionManager(config)
+
+    ###
+    # Methods that pass through to the connection manager
+    ###
+    # TODO: I would like a better way to do this. Should adapters be subclasses
```
Comment (drewbanin) [inline dbt/adapters/base/impl.py]: Can you say more about what a better version of this code looks like?

File: dbt/adapters/redshift/impl.py [hunk 3]
```diff
@@ -171,26 +95,9 @@ def drop_relation(self, relation, model_name=None):
...
-            self.begin(connection.name)
-
-            to_return = super(RedshiftAdapter, self).drop_relation(
-                relation, model_name)
-
-            self.commit(connection)
-            self.begin(connection.name)
-
-            return to_return
+        with self.connections.fresh_transaction(model_name):
```
Comment (drewbanin) [inline dbt/adapters/redshift/impl.py]: relevant: https://github.com/fishtown-analytics/dbt/issues/697

It looks like you're just using a brand new connection here, right? I _think_ that might be ok, but I'm curious what your thinking was around this change.

File: requirements.txt [hunk 4]
```diff
@@ -1,14 +1 @@
...
-minimal-snowplow-tracker==0.0.1
-snowflake-connector-python>=1.4.9
-colorama==0.3.9
-google-cloud-bigquery>=1.0.0,<2
-requests>=2.18.0,<3
-agate>=1.6,<2
-jsonschema==2.6.0
-boto3>=1.6.23,<1.8.0
-botocore>=1.9.23,<1.11.0
+-e .[postgres,snowflake,redshift,bigquery]
```
Comment (drewbanin) [inline requirements.txt]: does this just refer back to `setup.py`? That's really neat -- didn't know you could do that

File: setup.py [hunk 5]
```diff
@@ -39,20 +50,24 @@ def read(fname):
     scripts=[
         'scripts/dbt',
     ],
+    extras_require={
+        # when we split out adapters into their own python packages, these
+        # extras requirements lists will just be the appropriate adapter
+        # packages
+        'snowflake': ['snowflake-connector-python>=1.4.9'] + boto_requirements,
```
Comment (drewbanin) [inline setup.py]: HELL YEAH

File: dbt/adapters/base/connections.py (hunk 1 above)
Comment (beckjake) [reply dbt/adapters/base/connections.py]: This is actually not new to my code! I saw the `hasattr` check, thought "that's silly and can't be right", removed it, tested locally and found everything fine, and then appveyor failed.

So yeah, sometimes on windows the handle might not have a `close`, a fact I've determined empirically.

File: dbt/adapters/base/impl.py (hunk 2 above)
Comment (beckjake) [reply dbt/adapters/base/impl.py]: I should've removed this comment. I think it makes the class hierarchy very hard, but it does make using the classes nicer. Basically, the per-database connection managers would derive from the base connection manager and the base adapter would derive from the base connection manager as well. All adapter implementations would then derive from both the base adapter and their database-specific connection manager. Inheritance would be a mess but python's method resolution model would take care of it nicely. And we could remove the methods below this comment.

It's probably better to just preserve this small bit of ugliness.

File: dbt/adapters/redshift/impl.py (hunk 3 above)
Comment (beckjake) [reply dbt/adapters/redshift/impl.py]: This is functionally the same, I just extracted the transaction management into a context manager. It's not getting a new connection unless the named connection never existed. That context manager just triggers a (maybe commit) + begin, yields, and triggers a commit + begin again. Hence `fresh_transaction`.

I saw #697 and decided I'm not qualified on that front.

File: requirements.txt (hunk 4 above)
Comment (beckjake) [reply requirements.txt]: Yup! One of requirements.txt's nicest hidden features.

File: dbt/adapters/base/impl.py (hunk 2 above)
Comment (beckjake) [reply dbt/adapters/base/impl.py]:
```
BaseConnectionManager
               / \
              /   \
             /     \
   BaseAdapter    SQLConnectionManager
            \          /    \
             \        /      \
              \      /        \
             SQLAdapter    PostgresConnectionManager
                \                |
                 \   RedshiftConnectionManager
                  \    /                 
                 RedshiftAdapter
```

An ascii diagram of what (some of) the inheritance model would look like (this was so hard to make)

File: dbt/adapters/base/connections.py (hunk 1 above)
Comment (drewbanin) [reply dbt/adapters/base/connections.py]: i am at once shocked and not surprised :)

**Comment by beckjake** [comment]
Merged per our sprint planning discussion!

**Pull request merged**